cub/cub/device/dispatch/dispatch_transform.cuh

```diff
@@ -0,0 +1,995 @@
+      ::cuda::std::clamp(items_per_thread, +policy_t::MIN_ITEMS_PER_THREAD, +policy_t::MAX_ITEMS_PER_THREAD);
+    const Offset tile_size = block_dim * items_per_thread_clamped;
+    const auto grid_dim    = static_cast<unsigned int>(::cuda::ceil_div(num_items, tile_size));
+
+    return THRUST_NS_QUALIFIER::cuda_cub::launcher::triple_chevron(grid_dim, block_dim, 0, stream)
+      .doit(kernel,
+            num_items,
+            items_per_thread_clamped,
+            op,
+            out,
+            make_iterator_kernel_arg<THRUST_NS_QUALIFIER::try_unwrap_contiguous_iterator_t<RandomAccessIteratorsIn>>(
+              THRUST_NS_QUALIFIER::try_unwrap_contiguous_iterator(::cuda::std::get<Is>(in)))...);
+  }
+
+  template <std::size_t... Is>
+  struct non_contiguous_fallback_op_t
+  {
+    ::cuda::std::tuple<RandomAccessIteratorsIn...> in;
+    RandomAccessIteratorOut out;
+    mutable TransformOp op; // too many users forgot to mark there operator()'s const ...
+
+    _CCCL_HOST_DEVICE _CCCL_FORCEINLINE void operator()(Offset i) const
+    {
+      out[i] = op(::cuda::std::get<Is>(in)[i]...);
+    }
+  };
+
+  // TODO(bgruber): this algorithm is never dispatched to. It is just used by the benchmarks as a baseline for the first
+  // round of testing. Remove it later.
+  template <typename, std::size_t... Is>
+  CUB_RUNTIME_FUNCTION _CCCL_FORCEINLINE cudaError_t
+  invoke_algorithm(cuda::std::index_sequence<Is...>, ::cuda::std::integral_constant<Algorithm, Algorithm::fallback_for>)
+  {
+    using op_t = non_contiguous_fallback_op_t<Is...>;
+    return for_each::dispatch_t<Offset, op_t>::dispatch(num_items, op_t{in, out, op}, stream);
+  }
+
+  template <typename ActivePolicy>
+  CUB_RUNTIME_FUNCTION _CCCL_FORCEINLINE cudaError_t Invoke()
+  {
+    // // TODO(bgruber): replace the overload set by if constexpr in C++17
+    return invoke_algorithm<ActivePolicy>(::cuda::std::index_sequence_for<RandomAccessIteratorsIn...>{},
+                                          ::cuda::std::integral_constant<Algorithm, ActivePolicy::algorithm>{});
+  }
+
+  CUB_RUNTIME_FUNCTION _CCCL_FORCEINLINE static cudaError_t dispatch(
+    Offset num_items,
+    ::cuda::std::tuple<RandomAccessIteratorsIn...> in,
+    RandomAccessIteratorOut out,
+    TransformOp op,
+    cudaStream_t stream)
+  {
+    if (num_items == 0)
+    {
+      return cudaSuccess;
+    }
+
+    int ptx_version = 0;
+    auto error      = CubDebug(PtxVersion(ptx_version));
+    if (cudaSuccess != error)
+    {
+      return error;
+    }
+
+    // gevtushenko promised me that I can assume that stream belongs to the currently active device
+    int device = 0;
+    error      = CubDebug(cudaGetDevice(&device));
+    if (error != cudaSuccess)
+    {
+      return error;
+    }
+
+    int max_smem = 0;
+    error        = CubDebug(cudaDeviceGetAttribute(&max_smem, cudaDevAttrMaxSharedMemoryPerBlock, device));
+    if (error != cudaSuccess)
+    {
+      return error;
+    }
```

Comment:
suggestion: result of these queries is only needed in memcpy async and ublk kernels. Consider moving queries into corresponding launchers.

Reply:
Done.